### PR TITLE
hda: Hw hda chain xrun handling

### PR DESCRIPTION
**Pull request by makarukp**
This is draft which proposes xrun handling in chain dma. Please do not merge.

To enable that feature below PRs has to be merged:
https://github.com/thesofproject/sof/pull/6656
https://github.com/zephyrproject-rtos/zephyr/pull/53327

### CODE REVIEW

**Comment by lgirdwood** [comment]
@makarukp fyi - some conflicts so CI wont run.

File: src/audio/chain_dma.c [hunk 1]
```diff
@@ -47,6 +52,10 @@ struct chain_dma_data {
 	enum sof_ipc_stream_direction stream_direction;
 	/* container size in bytes */
 	uint8_t cs;
+#if CONFIG_IPC_MAJOR_4
+	bool under_over_run_notification_sent;
```
Comment (lyakh) [inline src/audio/chain_dma.c:56]: don't we usually call "onder-over-runs" just "xruns?"

File: src/audio/chain_dma.c [hunk 2]
```diff
@@ -224,6 +236,45 @@ static size_t chain_get_transferred_data_size(const uint32_t out_read_pos, const
...
+{
+	struct ipc4_resource_event_data_notification *notif_data = msg_xrun->tx_data;
+
+	notif_data->resource_id = resource_id;
+	notif_data->event_type = event_type;
+	notif_data->resource_type = GATEWAY;
+	notif_data->dws[0] = 0;
+	notif_data->dws[1] = 0;
+	notif_data->dws[2] = 0;
+	notif_data->dws[3] = 0;
```
Comment (lyakh) [inline src/audio/chain_dma.c:250]: `dws[]` has 6 elements, you set 4 of them to 0 here. Is the intention of this to make it deterministic? Do you need to zero out all of them? Use `memset()`?

File: src/audio/chain_dma.c [hunk 3]
```diff
@@ -243,6 +294,9 @@ static enum task_state chain_task_run(void *data)
 	case -EPIPE:
 		tr_warn(&chain_dma_tr, "chain_task_run(): dma_get_status() link xrun occurred,"
 			" ret = %u", ret);
+#if CONFIG_IPC_MAJOR_4
```
Comment (lyakh) [inline src/audio/chain_dma.c:297]: is chain DMA actually available with IPC3? Has it been tested with it? It seems IPC4-only to me, in fact we'd want a dependency somewhere in Kconfig I guess

File: src/include/ipc4/notification.h [hunk 4]
```diff
@@ -42,6 +42,17 @@ enum sof_ipc4_notification_type {
...
+	SOF_IPC4_GATEWAY_UNDERRUN_DETECTED	= 6,
+	SOF_IPC4_GATEWAY_OVERRUN_DETECTED	= 7,
+};
+
+/* Resource Type - source of the event */
+enum sof_ipc4_resource_type {
+	/* Gateway */
+	GATEWAY = 2,
+};
+
```
Comment (kv2019i) [inline src/include/ipc4/notification.h:55]: I guess these are generic definitions and could be a separate patch that adds the gateway resource event types to the common header. And then a second chain-dma specific patch that uses these flags for chain-dma.

File: src/audio/chain_dma.c [hunk 5]
```diff
@@ -224,6 +236,45 @@ static size_t chain_get_transferred_data_size(const uint32_t out_read_pos, const
...
+
+static void handle_xrun(struct chain_dma_data *cd)
+{
+	if ((cd->link_connector_node_id.f.dma_type == ipc4_hda_link_output_class)
+	    && (!cd->under_over_run_notification_sent)) {
+		tr_warn(&chain_dma_tr, "handle_xrun(): underrun detected");
+		xrun_notif_msg_init(cd->msg_xrun, cd->link_connector_node_id.dw,
+				    SOF_IPC4_GATEWAY_UNDERRUN_DETECTED);
+		ipc_msg_send(cd->msg_xrun, NULL, true);
+		cd->under_over_run_notification_sent = true;
```
Comment (kv2019i) [inline src/audio/chain_dma.c:261]: This probably should be good to put behind a Kconfig. At least with current Linux SOF driver, I don't see what we could do with these messages. Is driver supposed to react to these, or are these solely for debugging?

**Review by kv2019i — CHANGES_REQUESTED** [review]
A few minor notes. I think a bigger one is what is the purpose of this. We've not send xrun messages before for pipelines, so why send specifically for chain-dma...? Still a potentially good new tool, but something that needs clarification.

File: src/audio/chain_dma.c (hunk 5 above)
Comment (makarukp) [reply src/audio/chain_dma.c:261]: We should treat it as a useful information for debugging. Xrun notification message should be possible to be masked by driver (even in production fw), however we do not have this functionality right now in converge. In such case I will create new define in Kconfig for blocking xrun notifications.

File: src/audio/chain_dma.c [hunk 6]
```diff
@@ -589,6 +643,7 @@ static struct comp_dev *chain_task_create(const struct comp_driver *drv,
 	const uint32_t host_dma_id = cdma->primary.r.host_dma_id;
 	const uint32_t link_dma_id = cdma->primary.r.link_dma_id;
 	const uint32_t fifo_size = cdma->extension.r.fifo_size;
+	union ipc4_notification_header header;
```
Comment (mbukowsk) [inline src/audio/chain_dma.c:646]: Unused variable if CONFIG_IPC_MAJOR_4 is not defined

**Comment by makarukp** [comment]
To finalize that PR extra dependencies required:
#7023 
#7024

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by makarukp** [comment]
Please do not close, work in progress...

File: src/audio/chain_dma.c (hunk 5 above)
Comment (softwarecki) [reply src/audio/chain_dma.c:261]: I suggest move this function to some file under ipc/ipc4

Comment (makarukp) [reply src/audio/chain_dma.c:261]: @softwarecki I agree that 'xrun_notif_msg_init' could be used as generic function and it would be reasonable to move it to ipc/ipc4, however, I am not sure if 'handle_xrun' is good candidate as it is specific handling dedicated for chain dma interface.

**Comment by makarukp** [comment]
Requires #7164

File: src/include/ipc4/notification.h (hunk 4 above)
Comment (makarukp) [reply src/include/ipc4/notification.h]: @kv2019i Done in #7023

File: src/audio/chain_dma.c [hunk 7]
```diff
@@ -131,6 +140,34 @@ static size_t chain_get_transferred_data_size(const uint32_t out_read_pos, const
 	return buff_size - in_read_pos + out_read_pos;
 }
 
+#if CONFIG_XRUN_NOTIFICATIONS_ENABLE
```
Comment (lgirdwood) [inline src/audio/chain_dma.c:143]: This is an IPC4 specific code block, should the Kconfig not be ```CONFIG_IPC4_XRUN_NOTIFICATIONS_ENABLE```

**Review by kv2019i — APPROVED** [review]
Thanks, looks good now!

**Review by softwarecki — APPROVED** [review]

File: src/audio/chain_dma.c [hunk 8]
```diff
@@ -131,6 +140,34 @@ static size_t chain_get_transferred_data_size(const uint32_t out_read_pos, const
 	return buff_size - in_read_pos + out_read_pos;
 }
 
+#if CONFIG_IPC4_XRUN_NOTIFICATIONS_ENABLE
+static void handle_xrun(struct chain_dma_data *cd)
+{
+	int ret;
```
Comment (dnikodem) [inline src/audio/chain_dma.c:146]: unnecessary variable?

**Review by dnikodem — APPROVED** [review]
